src/ophyd_async/core/_soft_signal_backend.py

```diff
@@ -122,17 +127,46 @@ def make_initial_value(self, datatype: Optional[Type[T]]) -> T:
         return cast(T, self.choices[0])
 
 
+class SoftPydanticModelConverter(SoftConverter):
+    def __init__(self, datatype: Type[BaseModel]):
+        self.datatype = datatype
+
+    def reading(self, value: T, timestamp: float, severity: int) -> Reading:
+        value = self.value(value)
+        return super().reading(value, timestamp, severity)
+
+    def value(self, value: Any) -> Any:
+        if isinstance(value, dict):
+            value = self.datatype(**value)
+        return value
+
+    def write_value(self, value):
+        if isinstance(value, self.datatype):
+            return value.model_dump(mode="python")
+        return value
+
+    def make_initial_value(self, datatype: Type | None) -> Any:
+        return super().make_initial_value(datatype)
+
+
 def make_converter(datatype):
     is_array = get_dtype(datatype) is not None
     is_sequence = get_origin(datatype) == abc.Sequence
     is_enum = inspect.isclass(datatype) and (
         issubclass(datatype, Enum) or issubclass(datatype, RuntimeSubsetEnum)
     )
+    is_pydantic_model = (
+        inspect.isclass(datatype)
+        and isinstance(datatype, ABCMeta)
```

Comment:
If `datatype` is `numpy.ndarray[typing.Any, numpy.dtype[numpy.int8]]`
Then `inspect.isclass(datatype) is True`, `issubclass(datatype, Enum) is True`, and `issubclass(datatype, BaseModel)` gives

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/dls_sw/apps/python/miniforge/4.10.0-0/envs/python3.10/lib/python3.10/abc.py", line 123, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
TypeError: issubclass() arg 1 must be a class
```

This will be because abc metaclassing will do strange things with [`__subclasscheck__`](https://docs.python.org/3/reference/datamodel.html#class.__subclasscheck__), namely

https://github.com/python/cpython/blob/fb1b51a58df4315f7ef3171a5abeb74f132b0971/Lib/_py_abc.py#L108-L111
``` python
    def __subclasscheck__(cls, subclass):
        """Override for issubclass(subclass, cls)."""
        if not isinstance(subclass, type):
            raise TypeError('issubclass() arg 1 must be a class')
```

I was initially intrigued too, but it's a pain to debug and find out what's actually happening here with the hooks. That numpy type should pass this test.

Reply:
Ok, please could you put a comment in that says this is for npt.NDArray so we don't forget this oddity!